pkg/loop/plugin_service.go

```diff
@@ -0,0 +1,206 @@
+	utils.StartStopOnce
+
+	pluginName string
+
+	lggr logger.Logger
+	cmd  func() *exec.Cmd
+
+	wg     sync.WaitGroup
+	stopCh chan struct{}
+
+	grpcPlug P
+
+	client         *plugin.Client
+	clientProtocol plugin.ClientProtocol
+
+	newService func(context.Context, any) (S, error)
+
+	serviceCh chan struct{} // closed when service is available
+	service   S
+
+	testInterrupt chan func(*pluginService[P, S]) // tests only (via TestHook) to enable access to internals without racing
+}
+
+func newPluginService[P grpcPlugin, S types.Service](pluginName string, p P, newService func(context.Context, any) (S, error), lggr logger.Logger, cmd func() *exec.Cmd, stopCh chan struct{}) *pluginService[P, S] {
+	return &pluginService[P, S]{
+		pluginName: pluginName,
+		lggr:       lggr,
+		cmd:        cmd,
+		stopCh:     stopCh,
+		grpcPlug:   p,
+		newService: newService,
+		serviceCh:  make(chan struct{}),
+	}
+}
+
+func (r *pluginService[P, S]) keepAlive() {
+	defer r.wg.Done()
+
+	r.lggr.Debugw("Staring keepAlive", "tick", keepAliveTickDuration)
+
+	t := time.NewTicker(keepAliveTickDuration)
+	defer t.Stop()
+	for {
+		select {
+		case <-r.stopCh:
+			return
+		case <-t.C:
+			c := r.client
+			cp := r.clientProtocol
+			if c != nil && !c.Exited() && cp != nil {
+				// launched
+				err := cp.Ping()
+				if err == nil {
+					continue // healthy
+				}
+				r.lggr.Errorw("Relaunching unhealthy plugin", "err", err)
+			}
+			if err := r.tryLaunch(cp); err != nil {
+				r.lggr.Errorw("Failed to launch plugin", "err", err)
+			}
+		case fn := <-r.testInterrupt:
+			fn(r)
+		}
+	}
+}
+
+func (r *pluginService[P, S]) tryLaunch(old plugin.ClientProtocol) (err error) {
+	if old != nil && r.clientProtocol != old {
+		// already replaced by another routine
+		return nil
+	}
+	r.client, r.clientProtocol, err = r.launch()
+	return
+}
+
+func (r *pluginService[P, S]) launch() (*plugin.Client, plugin.ClientProtocol, error) {
+	ctx, cancelFn := utils.ContextFromChan(r.stopCh)
+	defer cancelFn()
+
+	r.lggr.Debug("Launching")
+
+	cc := r.grpcPlug.ClientConfig()
+	cc.Cmd = r.cmd()
+	client := plugin.NewClient(cc)
+	cp, err := client.Client()
+	if err != nil {
+		client.Kill()
+		return nil, nil, fmt.Errorf("failed to create plugin Client: %w", err)
+	}
+	abort := func() {
+		if cerr := cp.Close(); cerr != nil {
+			r.lggr.Errorw("Error closing ClientProtocol", "err", cerr)
+		}
+		client.Kill()
+	}
+	i, err := cp.Dispense(r.pluginName)
+	if err != nil {
+		abort()
+		return nil, nil, fmt.Errorf("failed to Dispense %q plugin: %w", r.pluginName, err)
+	}
+
+	select {
+	case <-r.serviceCh:
+		// already set
+	default:
+		r.service, err = r.newService(ctx, i)
+		if err != nil {
+			abort()
+			return nil, nil, fmt.Errorf("failed to create service: %w", err)
+		}
+		defer close(r.serviceCh)
+	}
+	return client, cp, nil
+}
+
+func (r *pluginService[P, S]) Start(context.Context) error {
+	return r.StartOnce("PluginService", func() error {
+		r.wg.Add(1)
+		go r.keepAlive()
```

Comment:
easier to read if the keepAlive accepts the wg as an arg

Reply:
Hmmm passing a struct field as a method param seems like an anti-pattern though... :thinking:

Reply:
It may be the only goroutine, but that may not always be the case, and this is a general, standard pattern used throughout our codebase, and go code in general. We could have `keepAlive` do the spawning, but I'm not sure that is better:
```go
func (r *pluginService[P, S]) keepAlive() {
	r.wg.Add(1)
	go func() {
		defer r.wg.Done()
		...
	}()
}
```
Because then we're hiding the `go` keyword, which seems like a worse version of the original issue.

Is this better?
```suggestion
		go func() {
			defer r.wg.Done()
			r.keepAlive()
		}()
```